Source/WebKit/UIProcess/XR/openxr/OpenXRLayer.cpp

```diff
@@ -110,6 +121,155 @@ std::optional<PlatformXR::FrameData::ExternalTexture> OpenXRLayer::exportOpenXRT
     };
 }
 
+#if USE(GBM)
+void OpenXRLayer::setGBMDevice(RefPtr<WebCore::GBMDevice> gbmDevice)
+{
+    m_gbmDevice = gbmDevice;
+}
+
+std::optional<PlatformXR::FrameData::ExternalTexture> OpenXRLayer::exportOpenXRTextureGBM(WebCore::GLDisplay& display, PlatformGLObject openxrTexture)
+{
+    auto preferredDMABufFormat = m_swapchain->format() == GL_RGBA8 ? DRM_FORMAT_ARGB8888 : DRM_FORMAT_XRGB8888;
+    WebCore::GLDisplay::DMABufFormat format;
+    const auto& supportedFormats = display.dmabufFormats();
+    for (const auto& supportedFormat : supportedFormats) {
+        if (supportedFormat.fourcc == preferredDMABufFormat) {
+            format = supportedFormat;
+            break;
+        }
+    }
+    if (!format.fourcc) {
+        RELEASE_LOG(XR, "OpenXR texture format not supported");
+        return std::nullopt;
+    }
+
+    auto* buffer = gbm_bo_create_with_modifiers2(m_gbmDevice->device(), m_swapchain->width(), m_swapchain->height(), format.fourcc, format.modifiers.span().data(), format.modifiers.size(), GBM_BO_USE_RENDERING);
+    if (!buffer)
+        buffer = gbm_bo_create(m_gbmDevice->device(), m_swapchain->width(), m_swapchain->height(), format.fourcc, GBM_BO_USE_RENDERING);
+    if (!buffer) {
+        RELEASE_LOG(XR, "Failed to allocate GBM buffer for OpenXR texture");
+        return std::nullopt;
+    }
+
+    Vector<UnixFileDescriptor> fds;
+    Vector<uint32_t> offsets;
+    Vector<uint32_t> strides;
+    uint32_t fourcc = gbm_bo_get_format(buffer);
+    uint64_t modifier = gbm_bo_get_modifier(buffer);
+    int planeCount = gbm_bo_get_plane_count(buffer);
+
+    Vector<EGLAttrib> attributes = {
+        EGL_WIDTH, static_cast<EGLAttrib>(gbm_bo_get_width(buffer)),
+        EGL_HEIGHT, static_cast<EGLAttrib>(gbm_bo_get_height(buffer)),
+        EGL_LINUX_DRM_FOURCC_EXT, static_cast<EGLAttrib>(fourcc),
+    };
+
+#define ADD_PLANE_ATTRIBUTES(planeIndex) { \
```

Comment:
This comment is out of scope for this PR and probably for a while, but I notice we have 7/8 definitions of `ADD_PLANE_ATTRIBUTES` that do a very similar thing but have some minor differences.

I wonder if there is a way to consolidate it elsewhere.